chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{error::ConnectTransactionError, TransactionSource};
+
+use common::{
+    chain::{Block, PoolId},
+    primitives::{Amount, Id},
+    Uint256,
+};
+use pos_accounting::{AccountingBlockRewardUndo, PoSAccountingOperations, PoSAccountingView};
+
+use super::accounting_delta_adapter::PoSAccountingDeltaAdapter;
+
+/// Distribute reward among the staker and delegators
+pub fn distribute_reward<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_reward: Amount,
+) -> Result<AccountingBlockRewardUndo, ConnectTransactionError> {
+    let tx_source = TransactionSource::Chain(block_id);
+
+    let pool_data = accounting_adapter
+        .accounting_delta()
+        .get_pool_data(pool_id)?
+        .ok_or(ConnectTransactionError::PoolDataNotFound(pool_id))?;
+
+    let total_staker_reward = (total_reward - pool_data.cost_per_block())
+        .and_then(|v| (v * pool_data.margin_ratio_per_thousand().into()).and_then(|v| v / 1000))
+        .and_then(|v| v + pool_data.cost_per_block())
+        .ok_or(ConnectTransactionError::StakerRewardCalculationFailed(
+            block_id,
+        ))?;
+    let total_delegators_reward = (total_reward - total_staker_reward).ok_or(
+        ConnectTransactionError::DelegatorRewardCalculationFailed(block_id),
+    )?;
+    let total_delegators_reward = Uint256::from_amount(total_delegators_reward);
```

Comment:
I do understand why you're doing this conversion, but I have to say that doing math operations with Uint256 is playing with fire, because none of the operations are checked for overflow/underflow problems.

Reply:
Multiplying uint128 shouldn't ever overflow uint256 and doing division should never underflow. So as long as we are doing only this set of operations we are safe, aren't we?

Reply:
Don't forget division by zero.

Generally speaking we ensure such mistakes won't happen by guarding operations with runtime checks. That's all I'm saying here. If someone decides to change the formula later, they may not notice the assumptions you made.

Reply:
Split into smaller functions and added check for zero devision